lightning_client.go

```diff
@@ -3564,3 +3574,86 @@ func (s *lightningClient) CheckMacaroonPermissions(ctx context.Context,
 
 	return res.Valid, nil
 }
+
+// InterceptFunction is a function type for intercepting RPC calls from a
+// middleware.
+type InterceptFunction func(context.Context,
+	*lnrpc.RPCMiddlewareRequest) (*lnrpc.RPCMiddlewareResponse, error)
+
+// RegisterRPCMiddleware adds a new gRPC middleware to the interceptor chain. A
+// gRPC middleware is software component external to lnd that aims to add
+// additional business logic to lnd by observing/intercepting/validating
+// incoming gRPC client requests and (if needed) replacing/overwriting outgoing
+// messages before they're sent to the client.
+func (s *lightningClient) RegisterRPCMiddleware(ctx context.Context,
+	middlewareName, customCaveatName string, readOnly bool,
+	timeout time.Duration, intercept InterceptFunction) (chan error, error) {
+
+	interceptStream, err := s.client.RegisterRPCMiddleware(
+		s.adminMac.WithMacaroonAuth(ctx),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We are expected to send an initial registration message immediately.
+	registerMsg := &lnrpc.RPCMiddlewareResponse{
+		MiddlewareMessage: &lnrpc.RPCMiddlewareResponse_Register{
+			Register: &lnrpc.MiddlewareRegistration{
+				MiddlewareName:           middlewareName,
+				CustomMacaroonCaveatName: customCaveatName,
+				ReadOnlyMode:             readOnly,
+			},
+		},
+	}
+	if err := interceptStream.SendMsg(registerMsg); err != nil {
+		return nil, err
+	}
+
+	errChan := make(chan error, 1)
+
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
Are we ok to only exit on caller context here, or do we need a quit channel? 

Asking because as is `WaitForFinished()` will not exit until the caller's context for `RegisterRPCMiddleware` is cancelled, even if we're shutting down lnd services.

With the current way we use `LndServices`, we always cancel our context for all calls before we call `Close`, but wondering whether it's apparent to an outside user that you must do so for this to exit?

Reply:
Hmm, good point. I've always looked at the `lndclient` instance as being more or less stateless so would expect the context needing to be canceled. Also I just copied this straight from the `ChannelAcceptor` code which behaves the same way.
Maybe we need to document this more clearly?
Looking at any of the other subscription methods, they all also only quit on context cancel.

Reply:
Yeah was just wondering about it, since we're just using cancel everywhere as is seems fine to continue with that pattern. Perhaps we can just drop a comment on the `Stop` method for lnd services that you have to cancel caller ctx to shutdown?